// fish-audio/dist/esm/api/resources/voices/Client.js
import * as environments from "../../../environments.js";
import * as core from "../../../core/index.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../core/headers.js";
import * as errors from "../../../errors/index.js";
import * as apiErrors from "../../../api/errors/index.js";
import { Ivc } from "./ivc/Client.js";
/**
 * Access to voices created either by you or Fish Audio.
 */
export class Voices {
    _options;
    _ivc;
    constructor(_options = {}) {
        this._options = _options;
    }
    get ivc() {
        return (this._ivc ??= new Ivc(this._options));
    }
    /**
     * Returns a list of all available voices for a user.
     *
     * @param {ModelListRequest} request
     * @param {Voices.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     await client.voices.search()
     */
    search(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__search(request, requestOptions));
    }
    async __search(request = {}, requestOptions) {
        const _queryParams = {};
        for (const [key, value] of Object.entries(request)) {
            if (value != null) {
                _queryParams[key] = value;
            }
        }
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, "model"),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling GET /model.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
    /**
     * Returns metadata about a specific voice.
     *
     * @param {string} voiceId - ID of the voice to be used.
     * @param {Voices.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     await client.voices.get("21m00Tcm4TlvDq8ikWAM")
     */
    get(voiceId, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get(voiceId, requestOptions));
    }
    async __get(voiceId, requestOptions) {
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, `model/${encodeURIComponent(voiceId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: requestOptions?.queryParams ?? {},
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling GET /model/{voice_id}.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
    /**
     * Deletes a voice by its ID.
     *
     * @param {string} voiceId - ID of the voice to be used.
     * @param {Voices.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     await client.voices.delete("21m00Tcm4TlvDq8ikWAM")
     */
    delete(voiceId, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__delete(voiceId, requestOptions));
    }
    async __delete(voiceId, requestOptions) {
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, `model/${encodeURIComponent(voiceId)}`),
            method: "DELETE",
            headers: _headers,
            queryParameters: requestOptions?.queryParams ?? {},
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling DELETE /model/{voice_id}.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
    /**
     * Edit a voice created by you.
     *
     * @param {string} voiceId
     * @param {UpdateModelRequest} request
     * @param {Voices.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.voices.update("21m00Tcm4TlvDq8ikWAM", {
     *         title: "title"
     *     })
     */
    update(voiceId, request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__update(voiceId, request, requestOptions));
    }
    async __update(voiceId, request, requestOptions) {
        const _request = await core.newFormData();
        if (request.title != null) {
            _request.append("title", request.title);
        }
        if (request.description != null) {
            _request.append("description", request.description);
        }
        if (request.cover_image != null) {
            await _request.appendFile("cover_image", request.cover_image);
        }
        if (request.visibility != null) {
            _request.append("visibility", request.visibility);
        }
        if (request.tags != null) {
            _request.append("tags", request.tags);
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, `model/${encodeURIComponent(voiceId)}`),
            method: "PATCH",
            headers: _headers,
            queryParameters: requestOptions?.queryParams ?? {},
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex ?? "half",
            body: _maybeEncodedRequest.body,
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling PATCH /model/{voice_id}.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
}
