// fish-audio/dist/cjs/core/fetcher/getResponseBody.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResponseBody = getResponseBody;
const BinaryResponse_js_1 = require("./BinaryResponse.js");
const ResponseWithBody_js_1 = require("./ResponseWithBody.js");
const json_js_1 = require("../json.js");
async function getResponseBody(response, responseType) {
    if (!(0, ResponseWithBody_js_1.isResponseWithBody)(response)) {
        return undefined;
    }
    switch (responseType) {
        case "binary-response":
            return (0, BinaryResponse_js_1.getBinaryResponse)(response);
        case "blob":
            return await response.blob();
        case "arrayBuffer":
            return await response.arrayBuffer();
        case "sse":
            return response.body;
        case "streaming":
            return response.body;
        case "text":
            return await response.text();
    }
    // if responseType is "json" or not specified, try to parse as JSON
    const text = await response.text();
    if (text.length > 0) {
        try {
            let responseBody = (0, json_js_1.fromJson)(text);
            return responseBody;
        }
        catch (err) {
            return {
                ok: false,
                error: {
                    reason: "non-json",
                    statusCode: response.status,
                    rawBody: text,
                },
            };
        }
    }
    return undefined;
}
