// fish-audio/dist/esm/api/resources/textToSpeech/Client.js
import * as environments from "../../../environments.js";
import * as core from "../../../core/index.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../core/headers.js";
import * as errors from "../../../errors/index.js";
import * as apiErrors from "../../../api/errors/index.js";
import { encode } from "@msgpack/msgpack";
export class TextToSpeech {
    _options;
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * Converts text into speech using a voice of your choice and returns audio.
     * @throws {@link apiErrors.UnprocessableEntityError}
     */
    convert(request, backend = "s1", requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__convert(backend, request, requestOptions));
    }
    async __convert(backend, request, requestOptions) {
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
            model: backend,
        }), requestOptions?.headers);
        //Serialize to msgpack
        let payload = request;
        if (Array.isArray(request.references)) {
            const refs = await Promise.all(request.references.map(async (r) => {
                const audio = r?.audio;
                if (typeof File !== "undefined" && audio instanceof File) {
                    const buf = new Uint8Array(await audio.arrayBuffer());
                    return { ...r, audio: buf };
                }
                return r;
            }));
            payload = { ...request, references: refs };
        }
        const body = encode(payload);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, `v1/tts`),
            method: "POST",
            headers: _headers,
            contentType: "application/msgpack",
            queryParameters: { ...requestOptions?.queryParams },
            requestType: "bytes",
            body: body,
            responseType: "streaming",
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling POST /v1/tts");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
}
