// fish-audio/dist/esm/api/resources/user/Client.js
import * as environments from "../../../environments.js";
import * as core from "../../../core/index.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../core/headers.js";
import * as errors from "../../../errors/index.js";
import * as apiErrors from "../../../api/errors/index.js";
export class User {
    _options;
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * Gets the user's API credit
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     await client.user.get_api_credit()
     */
    get_api_credit(requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get_api_credit(requestOptions));
    }
    async __get_api_credit(requestOptions) {
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, "wallet/self/api-credit"),
            method: "GET",
            headers: _headers,
            queryParameters: { ...requestOptions?.queryParams },
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling GET /wallet/self/api-credit.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
    /**
     * Gets the user's package
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     await client.user.get_package()
     */
    get_package(requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get_package(requestOptions));
    }
    async __get_package(requestOptions) {
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, "wallet/self/package"),
            method: "GET",
            headers: _headers,
            queryParameters: { ...requestOptions?.queryParams },
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling GET /wallet/self/package.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
}
