// fish-audio/dist/cjs/client/websocket/TextToSpeechWebSocketClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextToSpeechWebSocket = void 0;
const Client_js_1 = require("../../api/resources/textToSpeech/Client.js");
const ws_1 = __importDefault(require("ws"));
const connection_js_1 = require("./connection.js");
const core = __importStar(require("../../core/index.js"));
const environments = __importStar(require("../../environments.js"));
/**
 * Text-to-Speech WebSocket client.
 * @remarks
 * **Node.js only**: This class uses Node.js-specific APIs (WebSocket from 'ws').
 * It will not work in browsers, Deno, or Cloudflare Workers without modifications.
 */
class TextToSpeechWebSocket extends Client_js_1.TextToSpeech {
    constructor(options = {}) {
        super(options);
    }
    async getWebSocketUri() {
        // Get base URL from options, preferring baseUrl, then environment, then default Production
        const baseUrl = (await core.Supplier.get(this._options.baseUrl)) ??
            (await core.Supplier.get(this._options.environment)) ??
            environments.FishAudioEnvironment.Production;
        // Convert HTTP(S) to WS(S)
        const wsUrl = baseUrl.replace(/^https?:\/\//i, (match) => match.toLowerCase() === "https://" ? "wss://" : "ws://");
        return `${wsUrl}/v1/tts/live`;
    }
    /**
     * Establishes a WebSocket connection for real-time text-to-speech.
     *
     * @param options - Configuration options for the connection
     * @returns A promise that resolves to a RealtimeConnection instance
     *
     * @remarks
     * **Node.js only**: This method uses Node.js-specific APIs.
     *
     * @example
     * ```typescript
     * const connection = await client.textToSpeech.convertRealtime({
     *     request: TTSRequest,
     *     text_stream: "...",
     *     backend: "s1",
     * });
     * ```
     */
    async convertRealtime(request, textStream, backend = "s1") {
        let apiKey = this._options.apiKey;
        if (!apiKey) {
            throw new Error("API key is required");
        }
        // Resolve API key if it's a function or promise
        if (typeof apiKey === "function") {
            apiKey = apiKey();
        }
        if (apiKey instanceof Promise) {
            apiKey = await apiKey;
        }
        if (!apiKey) {
            throw new Error("API key is required");
        }
        const connection = new connection_js_1.RealtimeConnection();
        const uri = await this.getWebSocketUri();
        return new Promise((resolve) => {
            const websocket = new ws_1.default(uri, {
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    model: backend,
                },
            });
            // Attach websocket to connection immediately so error handlers are registered
            // This ensures errors during handshake (like 403) are properly emitted via event emitter
            connection.setWebSocket(websocket);
            // Resolve immediately with connection so users can attach listeners
            resolve(connection);
            websocket.on("open", () => {
                connection.send(new connection_js_1.StartEvent(request));
                (async () => {
                    try {
                        for await (const chunk of textStream) {
                            connection.send(new connection_js_1.TextEvent(chunk));
                        }
                    }
                    catch {
                        // swallow sender errors; receiver side will emit ERROR on finish=error
                    }
                    finally {
                        try {
                            connection.send(new connection_js_1.CloseEvent());
                        }
                        catch { }
                    }
                })();
            });
        });
    }
}
exports.TextToSpeechWebSocket = TextToSpeechWebSocket;
