// fish-audio/dist/cjs/client/FishAudioClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FishAudioClient = void 0;
const headers_js_1 = require("../core/headers.js");
const errors = __importStar(require("../errors/index.js"));
const TextToSpeechWebSocketClient_js_1 = require("./websocket/TextToSpeechWebSocketClient.js");
const Client_js_1 = require("../api/resources/voices/Client.js");
const Client_js_2 = require("../api/resources/speechToText/Client.js");
const Client_js_3 = require("../api/resources/user/Client.js");
class FishAudioClient {
    _options;
    _textToSpeech;
    _voices;
    _speechToText;
    _user;
    constructor(_options = {}) {
        const apiKey = _options.apiKey ?? process.env.FISH_API_KEY;
        if (apiKey == null) {
            throw new errors.FishAudioError({
                message: "Please pass in your FishAudio API Key or export FISH_API_KEY in your environment.",
            });
        }
        _options.apiKey = apiKey;
        this._options = {
            ..._options,
            headers: (0, headers_js_1.mergeHeaders)({
                "User-Agent": "fish-audio",
            }, typeof _options?.apiKey === "string" ? { Authorization: `Bearer ${_options.apiKey}` } : {}, _options?.headers),
        };
    }
    get textToSpeech() {
        return (this._textToSpeech ??= new TextToSpeechWebSocketClient_js_1.TextToSpeechWebSocket(this._options));
    }
    get voices() {
        return (this._voices ??= new Client_js_1.Voices(this._options));
    }
    get speechToText() {
        return (this._speechToText ??= new Client_js_2.SpeechToText(this._options));
    }
    get user() {
        return (this._user ??= new Client_js_3.User(this._options));
    }
}
exports.FishAudioClient = FishAudioClient;
