// fish-audio/dist/esm/core/form-data-utils/FormDataWrapper.js
import { toJson } from "../../core/json.js";
import { RUNTIME } from "../runtime/index.js";
function isNamedValue(value) {
    return typeof value === "object" && value != null && "name" in value;
}
function isPathedValue(value) {
    return typeof value === "object" && value != null && "path" in value;
}
function isStreamLike(value) {
    return typeof value === "object" && value != null && ("read" in value || "pipe" in value);
}
function isReadableStream(value) {
    return typeof value === "object" && value != null && "getReader" in value;
}
function isBuffer(value) {
    return typeof Buffer !== "undefined" && Buffer.isBuffer && Buffer.isBuffer(value);
}
function isArrayBufferView(value) {
    return ArrayBuffer.isView(value);
}
function getLastPathSegment(pathStr) {
    const lastForwardSlash = pathStr.lastIndexOf("/");
    const lastBackSlash = pathStr.lastIndexOf("\\");
    const lastSlashIndex = Math.max(lastForwardSlash, lastBackSlash);
    return lastSlashIndex >= 0 ? pathStr.substring(lastSlashIndex + 1) : pathStr;
}
async function streamToBuffer(stream) {
    if (RUNTIME.type === "node") {
        const { Readable } = await import("stream");
        if (stream instanceof Readable) {
            const chunks = [];
            for await (const chunk of stream) {
                chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
            }
            return Buffer.concat(chunks);
        }
    }
    if (isReadableStream(stream)) {
        const reader = stream.getReader();
        const chunks = [];
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                chunks.push(value);
            }
        }
        finally {
            reader.releaseLock();
        }
        const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
        const result = new Uint8Array(totalLength);
        let offset = 0;
        for (const chunk of chunks) {
            result.set(chunk, offset);
            offset += chunk.length;
        }
        return Buffer.from(result);
    }
    throw new Error("Unsupported stream type: " + typeof stream + ". Expected Node.js Readable stream or Web ReadableStream.");
}
export async function newFormData() {
    return new FormDataWrapper();
}
export class FormDataWrapper {
    fd = new FormData();
    async setup() {
        // noop
    }
    append(key, value) {
        this.fd.append(key, String(value));
    }
    getFileName(value, filename) {
        if (filename != null) {
            return filename;
        }
        if (isNamedValue(value)) {
            return value.name;
        }
        if (isPathedValue(value) && value.path) {
            return getLastPathSegment(value.path.toString());
        }
        return undefined;
    }
    async convertToBlob(value) {
        if (isStreamLike(value) || isReadableStream(value)) {
            const buffer = await streamToBuffer(value);
            const arrayBuffer = new ArrayBuffer(buffer.byteLength);
            new Uint8Array(arrayBuffer).set(buffer);
            return new Blob([arrayBuffer]);
        }
        if (value instanceof Blob) {
            return value;
        }
        if (isBuffer(value)) {
            const arrayBuffer = new ArrayBuffer(value.byteLength);
            new Uint8Array(arrayBuffer).set(value);
            return new Blob([arrayBuffer]);
        }
        if (value instanceof ArrayBuffer) {
            return new Blob([value]);
        }
        if (isArrayBufferView(value)) {
            const view = value;
            const source = new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
            const arrayBuffer = new ArrayBuffer(source.byteLength);
            new Uint8Array(arrayBuffer).set(source);
            return new Blob([arrayBuffer]);
        }
        if (typeof value === "string") {
            return new Blob([value]);
        }
        if (typeof value === "object" && value !== null) {
            return new Blob([toJson(value)], { type: "application/json" });
        }
        return new Blob([String(value)]);
    }
    async appendFile(key, value, fileName) {
        fileName = this.getFileName(value, fileName);
        const blob = await this.convertToBlob(value);
        if (fileName) {
            this.fd.append(key, blob, fileName);
        }
        else {
            this.fd.append(key, blob);
        }
    }
    getRequest() {
        return {
            body: this.fd,
            headers: {},
            duplex: "half",
        };
    }
}
