// fish-audio/dist/cjs/api/resources/speechToText/Client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeechToText = void 0;
const environments = __importStar(require("../../../environments.js"));
const core = __importStar(require("../../../core/index.js"));
const headers_js_1 = require("../../../core/headers.js");
const errors = __importStar(require("../../../errors/index.js"));
const apiErrors = __importStar(require("../../../api/errors/index.js"));
class SpeechToText {
    _options;
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * Transcribe an audio or video file.
     *
     * @param {STTRequest} request
     * @param {SpeechToText.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.speechToText.convert({
     *         audio: new File([], "audio.mp3"),
     *     })
     */
    convert(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__convert(request, requestOptions));
    }
    async __convert(request, requestOptions) {
        const _request = await core.newFormData();
        await _request.appendFile("audio", request.audio);
        if (request.language != null) {
            _request.append("language", request.language);
        }
        if (request.ignore_timestamps != null) {
            _request.append("ignore_timestamps", request.ignore_timestamps.toString());
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = (0, headers_js_1.mergeHeaders)(this._options?.headers, (0, headers_js_1.mergeOnlyDefinedHeaders)({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, "v1/asr"),
            method: "POST",
            headers: _headers,
            queryParameters: { ...requestOptions?.queryParams },
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex ?? "half",
            body: _maybeEncodedRequest.body,
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling POST /v1/asr.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
}
exports.SpeechToText = SpeechToText;
