// fish-audio/dist/esm/client/FishAudioClient.js
import { mergeHeaders } from "../core/headers.js";
import * as errors from "../errors/index.js";
import { TextToSpeechWebSocket } from "./websocket/TextToSpeechWebSocketClient.js";
import { Voices } from "../api/resources/voices/Client.js";
import { SpeechToText } from "../api/resources/speechToText/Client.js";
import { User } from "../api/resources/user/Client.js";
export class FishAudioClient {
    _options;
    _textToSpeech;
    _voices;
    _speechToText;
    _user;
    constructor(_options = {}) {
        const apiKey = _options.apiKey ?? process.env.FISH_API_KEY;
        if (apiKey == null) {
            throw new errors.FishAudioError({
                message: "Please pass in your FishAudio API Key or export FISH_API_KEY in your environment.",
            });
        }
        _options.apiKey = apiKey;
        this._options = {
            ..._options,
            headers: mergeHeaders({
                "User-Agent": "fish-audio",
            }, typeof _options?.apiKey === "string" ? { Authorization: `Bearer ${_options.apiKey}` } : {}, _options?.headers),
        };
    }
    get textToSpeech() {
        return (this._textToSpeech ??= new TextToSpeechWebSocket(this._options));
    }
    get voices() {
        return (this._voices ??= new Voices(this._options));
    }
    get speechToText() {
        return (this._speechToText ??= new SpeechToText(this._options));
    }
    get user() {
        return (this._user ??= new User(this._options));
    }
}
