// fish-audio/dist/esm/api/resources/voices/ivc/Client.js
import * as environments from "../../../../environments.js";
import * as core from "../../../../core/index.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as errors from "../../../../errors/index.js";
import * as apiErrors from "../../../../api/errors/index.js";
export class Ivc {
    _options;
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * Create a voice clone and add it to your Voices
     *
     * @param {ModelCreateRequest} request
     * @param {Ivc.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link apiErrors.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.voices.ivc.create({
     *         title: "title",
     *         voices: [fs.createReadStream("/path/to/your/file")],
     *     })
     */
    create(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__create(request, requestOptions));
    }
    async __create(request, requestOptions) {
        const _request = await core.newFormData();
        _request.append("type", request.type ?? "tts");
        _request.append("title", request.title);
        _request.append("train_mode", request.train_mode ?? "fast");
        for (const _file of request.voices) {
            await _request.appendFile("voices", _file);
        }
        if (request.visibility != null) {
            _request.append("visibility", request.visibility);
        }
        if (request.description != null) {
            _request.append("description", request.description);
        }
        if (request.cover_image != null) {
            await _request.appendFile("cover_image", request.cover_image);
        }
        if (request.texts != null) {
            _request.append("texts", request.texts);
        }
        if (request.tags != null) {
            _request.append("tags", request.tags);
        }
        if (request.enhance_audio_quality != null) {
            _request.append("enhance_audio_quality", request.enhance_audio_quality.toString());
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authHeader = requestOptions?.apiKey != null
            ? `Bearer ${requestOptions.apiKey}`
            : typeof this._options?.apiKey === "string"
                ? `Bearer ${this._options.apiKey}`
                : undefined;
        let _headers = mergeHeaders(this._options?.headers, mergeOnlyDefinedHeaders({
            Authorization: _authHeader,
        }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.FishAudioEnvironment.Production, "model"),
            method: "POST",
            headers: _headers,
            queryParameters: requestOptions?.queryParams ?? {},
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex ?? "half",
            body: _maybeEncodedRequest.body,
            timeoutMs: requestOptions?.timeoutInSeconds != null ? requestOptions.timeoutInSeconds * 1000 : 240000,
            ...(requestOptions?.maxRetries != null ? { maxRetries: requestOptions.maxRetries } : {}),
            ...(requestOptions?.abortSignal != null ? { abortSignal: requestOptions.abortSignal } : {}),
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new apiErrors.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.FishAudioError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        switch (_response.error.reason) {
            case "non-json":
                throw new errors.FishAudioError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.rawBody,
                    rawResponse: _response.rawResponse,
                });
            case "timeout":
                throw new errors.FishAudioTimeoutError("Timeout exceeded when calling POST /model.");
            case "unknown":
                throw new errors.FishAudioError({
                    message: _response.error.errorMessage,
                    rawResponse: _response.rawResponse,
                });
        }
    }
}
